### PR TITLE
FIX: When disagreeing with a flag that silenced a user, unsilence them

### PR DESCRIPTION
**Pull request by eviltrout**
Previously it would unhide their post but leave them silenced.

This fix also cleans up some of the helper classes to make it easier
to pass extra data to the silencing code (for example, a link to the
post that caused the user to be silenced.)

There used to be a SpamRulesEnforcer class, which took either a post or a
user as an argument, then depending on which one would call another class.
I refactored it to call those classes directly instead.

This patch also refactors the auto_silence specs to avoid using
stubs. I changed an internal method name and everything still passed, so 
that was a bad  omen.

### CODE REVIEW

File: app/services/spam_rule/auto_silence.rb [hunk 1]
```diff
@@ -1,37 +1,38 @@
...
+  def should_autosilence?
+    return false if @user.blank?
     return false if @user.staged?
     return false if @user.has_trust_level?(TrustLevel[1])
 
-    if SiteSetting.num_spam_flags_to_silence_new_user > (0) &&
-        SiteSetting.num_users_to_silence_new_user > (0) &&
-        num_spam_flags_against_user >= (SiteSetting.num_spam_flags_to_silence_new_user) &&
-        num_users_who_flagged_spam_against_user >= (SiteSetting.num_users_to_silence_new_user)
+    if SiteSetting.num_spam_flags_to_silence_new_user > 0 &&
```
Comment (eviltrout) [inline app/services/spam_rule/auto_silence.rb]: This is a minor style change. I don't know why you'd want to wrap the right side of a compare in parens.

Comment (ZogStriP) [reply app/services/spam_rule/auto_silence.rb]: Why didn't Rubocop pick that up?

File: app/services/spam_rule/flag_sockpuppets.rb [hunk 2]
```diff
@@ -5,11 +5,13 @@ def initialize(post)
   end
 
   def perform
-    if SiteSetting.flag_sockpuppets && reply_is_from_sockpuppet?
-      flag_sockpuppet_users
-      true
-    else
-      false
+    I18n.with_locale(SiteSetting.default_locale) do
```
Comment (ZogStriP) [inline app/services/spam_rule/flag_sockpuppets.rb]: 👍

File: app/services/user_silencer.rb [hunk 3]
```diff
@@ -16,6 +16,12 @@ def self.unsilence(user, by_user = nil, opts = {})
     UserSilencer.new(user, by_user, opts).unsilence
   end
 
+  def self.was_silenced_for?(post)
+    return false unless post.present?
```
Comment (ZogStriP) [inline app/services/user_silencer.rb]: Why not?

```suggestion
    return false if post.blank?
```

File: app/services/spam_rule/auto_silence.rb (hunk 1 above)
Comment (eviltrout) [reply app/services/spam_rule/auto_silence.rb]: Not sure, I'll take a look today and see if that's an option.

**Pull request merged**